test/test_crud_v2.py

```diff
@@ -0,0 +1,331 @@
+                    expected_ids = {}
+                    for str_index in ids:
+                        expected_ids[int(str_index)] = ids[str_index]
+                    if expected_ids != result.upserted_ids:
+                        return False
+                elif getattr(result, prop) != expected_result[res]:
+                    return False
+            return True
+        else:
+            if not expected_result:
+                return result is None
+            else:
+                return result == expected_result
+
+    def check_events(self, expected_events, listener):
+        res = listener.results
+        if not len(expected_events):
+            return
+
+        # Expectations only have CommandStartedEvents.
+        self.assertEqual(len(res['started']), len(expected_events))
+        for i, expectation in enumerate(expected_events):
+            event_type = next(iter(expectation))
+            event = res['started'][i]
+
+            # The tests substitute 42 for any number other than 0.
+            if (event.command_name == 'getMore'
+                    and event.command['getMore']):
+                event.command['getMore'] = 42
+            elif event.command_name == 'killCursors':
+                event.command['cursors'] = [42]
+            # Add upsert and multi fields back into expectations.
+            elif event.command_name == 'update':
+                updates = expectation[event_type]['command']['updates']
+                for update in updates:
+                    update.setdefault('upsert', False)
+                    update.setdefault('multi', False)
+
+            # Replace afterClusterTime: 42 with actual afterClusterTime.
+            expected_cmd = expectation[event_type]['command']
+            expected_read_concern = expected_cmd.get('readConcern')
+            if expected_read_concern is not None:
+                time = expected_read_concern.get('afterClusterTime')
+                if time == 42:
+                    actual_time = event.command.get(
+                        'readConcern', {}).get('afterClusterTime')
+                    if actual_time is not None:
+                        expected_read_concern['afterClusterTime'] = actual_time
+
+            for attr, expected in expectation[event_type].items():
+                actual = getattr(event, attr)
+                if isinstance(expected, dict):
+                    for key, val in expected.items():
+                        if val is None:
+                            if key in actual:
+                                self.fail("Unexpected key [%s] in %r" % (
+                                    key, actual))
+                        elif val == {'$exists': True}:
+                            if key not in actual:
+                                self.fail("Expected key [%s] in %r" % (
+                                    key, actual))
+                        elif val == {'$exists': False}:
```

Comment:
Jeremy put off adding `$exists` so this should be removed, see https://jira.mongodb.org/browse/SPEC-1215.